src/Database/LSMTree/Internal/RunReader.hs

```diff
@@ -187,10 +209,28 @@ next fs hbio reader@RunReader {..} = do
   Utilities
 -------------------------------------------------------------------------------}
 
+-- | Returns 'Nothing' on EOF.
+readDiskPageAt :: HasFS IO h -> Index.PageNo -> FS.Handle h -> IO (Maybe RawPage)
+readDiskPageAt fs pageNo h = do
+    mba <- newPinnedByteArray pageSize
+    -- TODO: make sure no other exception type can be thrown
+    handleJust (guard . FS.isFsErrorType FS.FsReachedEOF) (\_ -> pure Nothing) $ do
+      let byteOffset = pageNoToAbsOffset pageNo
+      _ <- FS.hGetBufExactlyAt fs h mba 0 (fromIntegral pageSize) byteOffset
+      ba <- unsafeFreezeByteArray mba
+      let !rawPage = unsafeMakeRawPage ba 0
+      return (Just rawPage)
+
+pageNoToAbsOffset :: Index.PageNo -> FS.AbsOffset
+pageNoToAbsOffset (Index.PageNo n) =
+    assert (n >= 0) $
+      FS.AbsOffset (mulPageSize (fromIntegral n))
+
 -- | Returns 'Nothing' on EOF.
 readDiskPage :: HasFS IO h -> FS.Handle h -> IO (Maybe RawPage)
 readDiskPage fs h = do
     mba <- newPinnedByteArray pageSize
+    -- TODO: make sure no other exception type can be thrown
```

Comment:
I don't think you can prevent this

Reply:
Yeah, I put that note in more as a reminder to review this area later.